Manage dependencies with PDM

vbaltrusaitis-reef opened the pull request:
This is a draft PR for two reasons:
* It depends on b2sdk fixing https://github.com/Backblaze/b2-sdk-python/issues/458 . A fix is in progress there. Currently the dependency points to my fork of b2sdk on GitHub.
* Before merging, Backblaze team needs to be notified of the changes in CI / `nox` so they can adjust for their internal tests.

Changes:
* pdm is used to manage dependencies, test and build `b2`.
* Docker image and bundles now use locked dependencies, providing consistency in tests vs publishing.
* All tests use locked dependencies. There is an unfortunate consequence - b2sdk is no longer installed from git master in CI. It would be possible to install b2sdk from master on top of locked dependencies but there are tradeoffs here:
  * If we use b2sdk from master for everything, there's no indication that CLI depends on unreleased features of b2sdk. If developed CLI features depend on new b2sdk features, we will need to make a commit that updates `pyproject.toml`/`pdm.lock` anyways before release. 
  * With pdm, it seems natural to `pdm add b2sdk @ git+https://github.com/Backblaze/b2-sdk-python` while developing a feature together with sdk.
  * If we're concerned with ensuring that new b2sdk are not somehow breaking the CLI, it might be better to resurrect https://github.com/Backblaze/b2-sdk-python/pull/260 
  * It would make sense to run tests on both latest release and master of b2sdk but that would blow up our already long-running workflow (I think tests are parallelized only up to some number of workers).

vbaltrusaitis-reef commented on .readthedocs.yml:
```diff
@@ -8,7 +8,11 @@ version: 2
+  jobs:
+    post_create_environment:
+      - pip install pdm
+      - pdm export --format requirements --prod --group doc --output requirements-doc.txt --no-hashes
```
If you have a non-PyPI dependency (like a dependency on a git repo), `pip install -r requirements.doc` will error out:
```
ERROR: Can't verify hashes for these requirements because we don't have a way to hash version control repositories:
    b2sdk@ git+https://github.com/vbaltrusaitis-reef/b2-sdk-python@ac1e1355f1aaac10a398618214021cc7ecb6a1eb from git+https://github.com/vbaltrusaitis-reef/b2-sdk-python@ac1e1355f1aaac10a398618214021cc7ecb6a1eb (from -r requirements-doc.txt (line 21))
```

To be fair, non-PyPI dependencies shouldn't get into master (and definitely definitely not release), so this is only a convenience for testing. But for this use-case this seems pretty harmless, I'm now on the fence if I should remove `--no-hashes`.

mjurbanski-reef replied:
don't think it matters much - this is run on readthedocs side afterall and only for docs generation;
so the worst case (i.e. someone hacked pypi depdency) our docs get defaced - but at that point it would very cheap price to pay for knowing one of our dependencies were hacked :D

vbaltrusaitis-reef commented on Dockerfile.template:
```diff
@@ -11,9 +11,21 @@ LABEL build-date-iso8601="${build_date}"
+ENV PDM_BUILD_SCM_VERSION=${version}
+RUN --mount=type=bind,source=./b2,target=/b2/b2 \
+    --mount=type=bind,source=./pyproject.toml,target=/b2/pyproject.toml \
+    --mount=type=bind,source=./pdm.lock,target=/b2/pdm.lock \
```
Yeah, lock is half the point. I did consider (and try out) a wheel+requirements approach: install using exported requirement file (`pdm export --no-self [...]`) and then install a `b2` wheel on top of that with `--no-deps`  - this would get rid of `pdm` within the container, less files to copy, it's nice. But in this case, I feel like I should definitely export requirements with hashes. However, this would fail the docker build when you have b2sdk from git as a dependency. And that makes for an unpleasant development experience in case you're trying out unreleased b2sdk features, you want to make a draft PR that uses them and see if all tests pass.

This was a bit of a dilemma, I feel that exported requirements + wheel is nicer in many ways.

mjurbanski-reef replied:
multi-layered dockerfilerfile is the way to go long term, not necessary now

How big image is after this change? our last b2 image was 170MB, so unless this one is much bigger (>250MB) then I feel we can easily try doing it later down the line.

When using requirements.txt I tried doing multilayer, but other than it being more error prone, the image size didn't differ much at all.

vbaltrusaitis-reef replied:
I checked the image size and yes, it increased by a lot after these changes, to 300Mb. So I implemented a multi-stage build and now it's 146M.

vbaltrusaitis-reef commented on noxfile.py:
```diff
@@ -659,6 +603,7 @@ def build_and_test_docker(session):
     """
     test_image_tag = 'b2:test'
     generate_dockerfile(session)
+    session.run('nox', '-s', 'dump_license', '-fb', 'venv', external=True, **run_kwargs)
```
AFAIK, there are three options:
1. Having `b2 license` not work in the docker image.
2. Installing `license` extras to docker image.
3. Pregenerating `b2/licenses_output.txt`.

This is one way to go with option 3 (which seems like the best option). I also considered multi-stage docker build (that would have other advantages too) but the docker image definition was getting a little convoluted that way.

vbaltrusaitis-reef commented on pyproject.toml:
```diff
@@ -19,18 +21,43 @@ classifiers = [
+dependencies = [
+    "argcomplete>=2,<4",
+    "arrow>=1.0.2,<2.0.0",
+    "b2sdk @ git+https://github.com/vbaltrusaitis-reef/b2-sdk-python",
```
Yes, absolutely. I'll change this before merging and submit for a review once b2sdk SyntaxWarning problem is fixed.

`nox -s make_release_commit` would detect this dependency and complain. That's, unfortunately, not the same as preventing this from reaching master, but it's something.

mjurbanski-reef commented on Dockerfile.template:
```diff
@@ -11,9 +11,21 @@ LABEL build-date-iso8601="${build_date}"
-RUN --mount=type=bind,source=${tar_path}/${tar_name},target=/tmp/${tar_name} \
-  pip install --no-cache-dir /tmp/${tar_name}[full] && rm -rf /root/.cache
+# In case we're installing from git repo
+RUN apt-get update -y && apt-get install git -y
+
+RUN pip install -U pdm
+
+ENV PDM_BUILD_SCM_VERSION=${version}
+RUN --mount=type=bind,source=./b2,target=/b2/b2 \
```
if you combine there three commands and at the end remove extra stuff,
i.e.
git (I think apt-get purge will remove unneeded deps as well if there are any)
pdm (idk how to remove its dependencies, but lets not worry about it)
and apt get cache

it will noticeably lower the image size at relatively low cost of not-so-nice looking dockerfile

Until we have mutli-stage docker image IMO that would be better.

Personally I kinda feel like installing git here is not necessary and not relevant to PDM change, but I guess we can come back to that discussion later as I hope it won't influence image size much after proposed tweak.

vbaltrusaitis-reef replied:
Instead of this, I implemented multi-stage build, as suggested in the comment above. Building may still be wasteful but the end result is fine.

mjurbanski-reef replied:
love the multi-stage build <3 seems awesome (and not wasteful IMO, more layers in builder allow for better caching)

vbaltrusaitis-reef commented:
In the last few commits:
- Implemented multi-stage Docker build (as per https://pdm-project.org/latest/usage/advanced/#use-pdm-in-a-multi-stage-dockerfile).
- Moved `license-outputs.txt` generation to within Docker build.
- Updated `pyproject.toml` to point to `b2sdk` from PyPI instead of local repo.
- Fixed some issues where pyinstaller binaries would show `b2 version` as `<version>.editable`.

Requesting re-review @mjurbanski-reef

mjurbanski-reef commented on changelog.d/+pdm.infrastructure.md:
```diff
@@ -0,0 +1 @@
+Use pdm for building, testing and managing dependencies.
```
I think seeing at issue reported in other repo it is worth mentioning here that tests and other superfluous files have been removed from sdist tarball.
